src/Build/BackEnd/Components/Scheduler/Scheduler.cs

```diff
@@ -1781,6 +1786,66 @@ private ScheduleResponse TrySatisfyRequestFromCache(int nodeForResults, BuildReq
             return null;
         }
 
+        private bool CacheMissEnforcement(int nodeForResults, BuildRequest request, List<ScheduleResponse> responses)
+        {
+            var isIsolatedBuild = _componentHost.BuildParameters.IsolateProjects;
+
+            // do not check root requests as nothing depends on them
+            if (!request.IsRootRequest && isIsolatedBuild)
+            {
+                var configCache = (IConfigCache)_componentHost.GetComponent(BuildComponentType.ConfigCache);
+                var requestConfig = configCache[request.ConfigurationId];
+                
+                // Need the parent request. But the parent / child relationship is sometimes formed after the cache is queried, so get creative
+                var parentRequest = _schedulingData.BlockedRequests.FirstOrDefault(r => r.BuildRequest.GlobalRequestId == request.ParentGlobalRequestId);
+                if (parentRequest == null)
+                {
+                    // the parent might still be in executing requests because the scheduler might not have had a chance to mark it as blocked yet
```

Comment:
Based on the comments, do we need to worry about thread-safety here? And/or a race condition between `BlockedRequests` and `ExecutingRequests`?

Reply:
I don't think so. The comment is trying to say that the scheduler sometimes checks the cache before it constructs the parent - child relationship (these are sequential steps). And in addition, when the scheduler checks the cache, there cannot be a concurrent insertion with the same data that would lead to a sort of read-before-write race.